File: java/jzoffer/chapter5/IntersectionNode_52.java
package chapter5;

public class IntersectionNode_52 {
    public class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }

    public int length(ListNode head) {
        int counter = 0;
        while (head != null) {
            counter++;
            head = head.next;
        }
        return counter;
    }

    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int lengthA = length(headA), lengthB = length(headB);
        if (lengthB > lengthA) return getIntersectionNode(headB, headA);

        int diff = lengthA - lengthB;
        while (diff > 0) {
            diff--;
            headA = headA.next;
        }

        while (headA != headB) {
            headA = headA.next;
            headB = headB.next;
        }
        return headA;
    }
}

File: go/jzoffer/intersectionNode_52.go
package jzoffer

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	lenA, lenB := getLen(headA), getLen(headB)
	if lenA < lenB {
		return getIntersectionNodeWithLen(headB, headA, lenB, lenA)
	}
	return getIntersectionNodeWithLen(headA, headB, lenA, lenB)
}

func getIntersectionNodeWithLen(headA, headB *ListNode, lenA, lenB int) *ListNode {
	diff := lenA - lenB

	for diff > 0 {
		diff--
		headA = headA.Next
	}

	for headA != headB {
		if headA == nil || headB == nil {
			return nil
		}
		headA = headA.Next
		headB = headB.Next
	}
	return headA
}

func getLen(node *ListNode) int {
	res := 0

	for node != nil {
		res++
		node = node.Next
	}
	return res
}
